Best practices guide on DAG writing on Astro (#3797)
astronomer/docs: astro/best-practices/dag-writing-on-astro.md

---

> title: 'Optimize Airflow DAGs for Astro'

Reviewer: Why do we have the title as "optimize"? I would keep it the same as teh sidebar "DAG writing on Astro" unless there was a specific reason we chose this

sidebar_label: 'DAG writing on Astro'
id: dag-writing-on-astro
---

You can run any valid Airflow DAG on Astro. 

That said, following best practices and taking full advantage of Astro features will help you:

- Produce DAGs more efficiently.
- Improve your development workflow.
- Better organize your deployment and DAGs. 
- Optimize resource usage and task execution efficiency.

> While this best practice guide focuses on writing Airflow DAGs for Astro, all [general Airflow best practices](https://docs.astronomer.io/learn/dag-best-practices) are also recommended.

Reviewer: ```suggestion
This guide describes best practices for taking advantage of Astro features when writing DAGs. While this focuses on writing Airflow DAGs for Astro, all [general Airflow best practices](https://docs.astronomer.io/learn/dag-best-practices) are also recommended.
```

:::info New to Airflow?

If you are new to Airflow, we advise starting with the following resources:

- Hands-on tutorial: [Get started with Apache Airflow](https://docs.astronomer.io/learn/get-started-with-airflow).
- Astronomer Academy: [Airflow 101 Learning Path](https://academy.astronomer.io/path/airflow-101).
- Webinar: [Airflow 101: How to get started writing data pipelines with Apache Airflow](https://www.astronomer.io/events/webinars/airflow-101-how-to-get-started-writing-data-pipelines-with-apache-airflow-video/).

:::

## Feature overview

In this guide, you'll learn about a number of Astro and Airflow features for optimizing your DAGs for Astro:

- [Astro Cloud IDE](https://docs.astronomer.io/astro/cloud-ide). A low-code, notebook-inspired IDE for writing and testing pipelines on Astro.
- [Astro CLI](overview.md). An OSS tool for developing DAGs and deploying projects tailor-made for Astro.

> - [CI/CD on Astro](set-up-ci-cd.md). A recommended architecture for testing and deployment supported by pre-built templates.

Reviewer: ```suggestion
```

- [Worker queues](configure-worker-queues.mdx). A feature of Astro that enables resource optimization beyond what is possible using OSS Airflow.
- [Astro Cloud UI Environment Manager](manage-connections-variables.md). A tool for managing connections across multiple deployments locally and on Astro.
- [Astro alerts](alerts.md). No-code alerting on Astro configurable for various trigger types and communication channels.

> - [Kubernetes Executor](executors-overview.md). Supports running Airflow on dedicated Kubernetes pods on Astro.
> - [KubernetesPodOperator](kubernetespodoperator.md). Supports running specific tasks in dedicated Kubernetes pods on Astro.

Reviewer: ```suggestion
```

> ## Best practice guidance

Reviewer: I commented below on removing a few sections in here. Overall I think the ones we need to keep are on development (Cloud IDE, CLI), and Astro features that would change DAG code (what is currently in Leverage Astro resources for improved Airflow management, maybe we call this "DAG code on Astro" or something like that).

So, let's have two sub-sections in here, one for each of those sets of guidance. For development, the sections on Cloud IDE and CLI are good content-wise, but I would add a few sentences on when you might choose one or the other (or both). 

For the Leverage Astro resources for improved Airflow management stuff, we should make clear that these are features that require DAG code changes (or in the case of connections, adding config in Astro instead of Airflow), and describe why somebody would choose those. When we say "your DAGs will just run on Astro", we should frame this section as "here are some changes you may consider making when migrating or writing new DAGs"

### Use the Astro CLI for local development

When writing DAGs intended for Astro Deployments, use the [Astro CLI](https://docs.astronomer.io/astro/cli/install-cli) for containerized local development. The Astro CLI is an open-source interface you can use to:
- test Airflow DAGs locally
- deploy code to Astro
- automate key actions as part of a CI/CD process. 

To get the most out of the Astro CLI: 

- Use the `dags` directory to store your DAGs. If you have multiple DAGs, you can use subdirectories to keep your environment organized.
- Modularize your code and store all supporting classes, files and functions in the `include` directory.
- Store all your [tests](#follow-devops-best-practices) in the `tests` directory.

:::tip

If you are unable to install the Astro CLI on your local machine, due to company policy or for other reasons, you can use it in GitHub Codespaces by forking the [Astro CLI Codespaces](https://github.com/astronomer/astro-cli-codespaces) repository.

:::

Learn more and get started by following the guidance in [Astro CLI](https://www.astronomer.io/docs/astro/cli/overview).

> ### Follow DevOps best practices

Reviewer: Let's remove this section. I think we should keep this guide to only best practices for writing DAGs (what tools to use for writing them, and what should go in the code that might be different from OSS Airflow), not also deploying them (we have deployment stuff in other best practices and can always write more).

Test and validate your Airflow pipelines prior to deployment:

- Write **DAG validation tests** to create rules to, for example, enforce the use of specific operators or allow only certain schedules. Store all validation tests in the `tests` directory.
- Write **unit tests** for all your custom code using your preferred Python testing framework. Popular options include `pytest` and `unittest`. Store all unit tests in the `tests` directory.
- Consider writing **integration tests** for custom functions that connect to external tools and store them in the `tests` directory, as well. A Python testing framework can also be used for this purpose.

You can run all tests stored in the `tests` directory with the `astro dev pytest` command and integrate this command into your CI/CD pipelines. For more information about testing Airflow DAGs, see [Test Airflow DAGs](https://docs.astronomer.io/learn/testing-airflow).

> ### Follow DAG-writing best practices

Reviewer: None of this section is Astro related, we can remove

- **[Deployment environment variables](environment-variables.md)** for storing environment variables. You can mark variables as secret for storage in the Astronomer managed secrets backend. You can store [Airflow variables](https://docs.astronomer.io/learn/airflow-variables) as environment variables as well by using the format `AIRFLOW_VAR_MYVARIABLENAME`.
- **[Worker queues](https://docs.astronomer.io/astro/configure-worker-queues)** for optimizing task execution efficiency. Worker queues allow you to define sets of workers with specific resources that execute tasks in parallel. You can assign tasks to specific worker queues to optimize resource usage. This feature is an improvement over OSS Airflow pools, which only manage concurrency and not resources. You can use both worker queues and pools in combination.
- **[Astro Alerts](alerts.md)** for setting up alerts for your DAGs. For more information, see [When to use Airflow or Astro alerts for your pipelines on Astro](airflow-vs-astro-alerts.md).

> ### Kubernetes and Astro

Reviewer: Similar to above comment, let's axe this section as well

On Astro, Airflow runs on Kubernetes. To run Airflow tasks in dedicated Kubernetes pods you have two options:

- **Kubernetes Executor**: on Astro you can switch to the Kubernetes Executor in the deployment settings. This will run all tasks running in a deployment in dedicated Kubernetes pods. See [Manage Airflow executors on Astro](executors-overview.md).
- **KubernetesPodOperator**: if you are using the Celery Executor, you can use the KubernetesPodOperator to run specific tasks in dedicated Kubernetes pods. You can run your tasks on the same cluster as your Airflow deployment by using the `in_cluster` parameter, or on a different cluster by using the `cluster_context` parameter. See [Run the KubernetesPodOperator on Astro](kubernetespodoperator.md).

> ## Astro DAG Example 

Reviewer: I think we can axe the whole example - all of the features we're talking about here (once we remove some stuff) are straightforward to implement and clearly described in their respective docs.

Making use of the following features can require DAG code changes or adding config in Astro instead of Airflow. Astronomer recommends considering these features during DAG development to streamline the development process and make use of everything Astro has to offer.

- **[Astro Cloud UI Environment Manager](https://docs.astronomer.io/astro/manage-connections-variables)** for managing connections. This allows you to define connections once and use them in multiple deployments (with or without overrides for individual fields), as well as in your local development environment. Connections in the Environment Manager are stored in the Astronomer managed secrets backend. Using the Environment Manager requires adding packages to your Astro project and adding config in Astro.

> - **[Deployment environment variables in the Astro UI](environment-variables.md)** for storing environment variables. Common uses of environment variables include adding tokens or URLs required by your DAGs, integrating with third-party tooling to export metrics, customizing core settings of Airflow or Airflow Providers, and storing Airflow connections and variables. A number of approaches are supported. For example, you can mark variables as secret for storage in the Astronomer managed secrets backend. You can store [Airflow variables](https://docs.astronomer.io/learn/airflow-variables) as environment variables as well by using the format `AIRFLOW_VAR_MYVARIABLENAME`. You can also manage environment variables using your Astro project Dockerfile and, locally, your project `.env` file. Astro project modification and additional config in Astro may be required depending on the strategy chosen.

Reviewer 2: ```suggestion
- **[Deployment environment variables in the Astro UI](environment-variables.md)** for storing environment variables. Common uses of environment variables include adding tokens or URLs required by your DAGs, integrating with third-party tooling to export metrics, customizing core settings of Airflow or Airflow Providers, and storing Airflow connections and variables. A number of approaches are supported. For example, you can mark variables as secret for storage in the Astronomer managed secrets backend. You can store [Airflow variables](https://docs.astronomer.io/learn/airflow-variables), which are Airflow native key-value pairs, as environment variables as well by using the format `AIRFLOW_VAR_MYVARIABLENAME`. You can also manage environment variables using your Astro project Dockerfile and, locally, your project `.env` file. Astro project modification and additional configuration in Astro may be required depending on the strategy chosen.
```

Another clarification suggestion in case users very new to Airflow read this to hopefully make the concept of Airflow variables more distinct from all other kinds of variables 😅

#### Use the Astro CLI for local development

When writing DAGs intended for Astro Deployments, use the [Astro CLI](https://docs.astronomer.io/astro/cli/install-cli) for containerized local development. The Astro CLI is an open-source interface you can use to:
- test Airflow DAGs locally
- deploy code to Astro
- automate key actions as part of a CI/CD process. 

To get the most out of the Astro CLI: 

- Use the `dags` directory to store your DAGs. If you have multiple DAGs, you can use subdirectories to keep your environment organized.
- Modularize your code and store all supporting classes, files and functions in the `include` directory.
- Store all your [tests](#follow-devops-best-practices) in the `tests` directory.

:::tip

If you are unable to install the Astro CLI on your local machine, due to company policy or for other reasons, you can use it in GitHub Codespaces by forking the [Astro CLI Codespaces](https://github.com/astronomer/astro-cli-codespaces) repository.

:::

Learn more and get started by following the guidance in [Astro CLI](https://www.astronomer.io/docs/astro/cli/overview).

### Managing DAG code on Astro

> Compared to open-source Airflow, Astro offers a distinct feature set for managing deployments that you should consider not only when writing new DAG code but also when migrating DAGs to Astro. Although it is true that your DAGs will just run when you migrate them, you also get access to an upgraded experience when it comes to managing connections, defining and storing environment variables, allocating resources, and implementing alerts.

Reviewer: ```suggestion
Compared to open-source Airflow, Astro offers a distinct feature set for optimizing your Airflow use. Some of these features impact individual DAGs, and you should consider them not only when writing new DAG code but also when migrating DAGs to Astro. You don't need to make changes to existing DAGs when you migrate them, but by implementing some small changes you can get access to an upgraded experience when it comes to managing connections, defining and storing environment variables, allocating resources, and implementing alerts.
```

> ::: note

Reviewer: ```suggestion
```

> Making use of these features can require DAG code changes or adding config in Astro instead of Airflow.

Reviewer: ```suggestion
Making use of the following features can require DAG code changes or adding config in Astro instead of Airflow. Astronomer recommends considering these features during DAG development to streamline the development process and make use of everything Astro has to offer.
```

> :::

Reviewer: ```suggestion
```

> - **[Astro Cloud UI Environment Manager](https://docs.astronomer.io/astro/manage-connections-variables)** for managing connections. Allows you to define connections once and use them in multiple deployments (with or without overrides for individual fields), as well as in your local development environment. Connections in the Environment Manager are stored in the Astronomer managed secrets backend. Requires adding packages to your Astro project and adding config in Astro.

Reviewer: ```suggestion
- **[Astro Cloud UI Environment Manager](https://docs.astronomer.io/astro/manage-connections-variables)** for managing connections. This allows you to define connections once and use them in multiple deployments (with or without overrides for individual fields), as well as in your local development environment. Connections in the Environment Manager are stored in the Astronomer managed secrets backend. Using the Environment Manager requires adding packages to your Astro project and adding config in Astro.
```
Which packages are required?

> - **[Deployment environment variables](environment-variables.md)** for storing environment variables. A number of approaches are supported. For example, you can mark variables as secret for storage in the Astronomer managed secrets backend. You can store [Airflow variables](https://docs.astronomer.io/learn/airflow-variables) as environment variables as well by using the format `AIRFLOW_VAR_MYVARIABLENAME`. Astro project modification and additional config in Astro possibly required depending on the strategy chosen.

Reviewer: I think we should add something in here about why you would want to use deployment environment variables over other ways of managing them.

> - **[Worker queues](https://docs.astronomer.io/astro/configure-worker-queues)** for optimizing task execution efficiency. Worker queues allow you to define sets of workers with specific resources that execute tasks in parallel. You can assign tasks to specific worker queues to optimize resource usage. This feature is an improvement over OSS Airflow pools, which only manage concurrency and not resources. You can use both worker queues and pools in combination. Config in Astro and DAG code modification required.

Reviewer: ```suggestion
- **[Worker queues](https://docs.astronomer.io/astro/configure-worker-queues)** for optimizing task execution efficiency. Worker queues allow you to define sets of workers with specific resources that execute tasks in parallel. You can assign tasks to specific worker queues to optimize resource usage. This feature is an improvement over OSS Airflow pools, which only manage concurrency and not resources. You can use both worker queues and pools in combination. Config in Astro and DAG code modification are required to use this feature.
```

> - **[Astro Alerts](alerts.md)** for setting up alerts for your DAGs. For more information, see [When to use Airflow or Astro alerts for your pipelines on Astro](airflow-vs-astro-alerts.md). Config in Astro required.

Reviewer: Here let's also add something brief on why you might want to use them - and then can link out to the other best practice guide for more info (as you do now)
